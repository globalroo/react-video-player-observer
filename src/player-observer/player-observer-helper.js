export const PLAYER_ERRORS = {
	DECODING_ERROR: "Decoding error",
	FETCH_ABORTED: "Fetch aborted",
	NETWORK_ERROR: "Network error",
	STATE_OK: "OK",
	UNKNOWN_ERROR: "Unknown error",
	UNSUITABLE_MEDIA: "Unsuitable media",
};

export const getPlayerError = ({ code = 0 }) => {
	let description;
	switch (code) {
	case 0:
		description = PLAYER_ERRORS.STATE_OK;
		break;
	case 1:
		description = PLAYER_ERRORS.FETCH_ABORTED;
		break;
	case 2:
		description = PLAYER_ERRORS.NETWORK_ERROR;
		break;
	case 3:
		description = PLAYER_ERRORS.DECODING_ERROR;
		break;
	case 4:
		description = PLAYER_ERRORS.UNSUITABLE_MEDIA;
		break;
	default:
		description = PLAYER_ERRORS.UNKNOWN_ERROR;
	}
	return {
		code,
		description,
	};
};

export const getRemainingTime = ({ duration = 0, currentTime = 0 }) => duration - currentTime;

export const getObservablePlayerProperties = (player, evt) => ({
	buffered: player.buffered,
	currentSrc: player.currentSrc,
	currentTime: player.currentTime,
	duration: player.duration,
	error: getPlayerError(player.error || {}),
	nativeEvent: evt,
	paused: player.paused,
	remainingTime: getRemainingTime(player),
	seekeable: player.seekable,
	seeking: player.seeking,
	src: player.src,
});
