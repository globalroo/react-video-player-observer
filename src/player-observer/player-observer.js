import { getObservablePlayerProperties } from "src/player-observer/player-observer-helper";

export const createDispatcher = (context, eventName) => (evt) => context.dispatch(eventName, evt);

export class PlayerObserver {
	static instance = null;

	static initialised() {
		return (PlayerObserver.instance) ? true : false;
	}

	static getInstance() {
		if (!PlayerObserver.initialised()) {
			PlayerObserver.instance = new PlayerObserver();
		}
		return PlayerObserver.instance;
	}

	constructor() {
		this.events = {};
	}

	initialisePlayer(player) {
		this.player = player;
	}

	bindEvent(eventName) {
		this.events[eventName] = {};
		this.events[eventName].subscribers = [];
		const dispatcher = createDispatcher(this, eventName);
		this.player.addEventListener(eventName, dispatcher);
	}

	dispatch(eventName, evt) {
		if (!this.events[eventName]) {
			return -1;
		}
		this.events[eventName].subscribers.forEach(
			(subscriber) =>
				subscriber(eventName, getObservablePlayerProperties(this.player, evt))
		);
	}

	subscribe(eventName, subscriber) {
		if (!this.events[eventName]) {
			this.bindEvent(eventName);
		}
		this.events[eventName].subscribers.push(subscriber);
		return () => {
			this.events[eventName].subscribers = this.events[eventName].subscribers.filter((fn) => fn !== subscriber);
		};
	}
}
