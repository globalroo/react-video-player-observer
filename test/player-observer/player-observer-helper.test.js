import { getPlayerError, getObservablePlayerProperties } from "src/player-observer/player-observer-helper";

describe ("Player observer helper test", () => {
	test("Error code 0 returns as expected", () => {
		const expected = {"code": 0, "description": "OK"};
		const code = 0;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
	test("Error code 1 returns as expected", () => {
		const expected = {"code": 1, "description": "Fetch aborted"};
		const code = 1;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
	test("Error code 2 returns as expected", () => {
		const expected = {"code": 2, "description": "Network error"};
		const code = 2;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
	test("Error code 3 returns as expected", () => {
		const expected = {"code": 3, "description": "Decoding error"};
		const code = 3;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
	test("Error code 4 returns as expected", () => {
		const expected = {"code": 4, "description": "Unsuitable media"};
		const code = 4;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
	test("Error code unknown returns as expected", () => {
		const expected = {"code": 12345, "description": "Unknown error"};
		const code = 12345;
		const errorObj = getPlayerError({ code });
		expect(errorObj).toEqual(expected);
	});
});

describe ("getObservableProperties returns expected structure", () => {
	test("getObservablePlayerProperties", () => {
		const evt = "test";
		const player = {
			buffered: 0,
			currentSrc: "test",
			currentTime: 10,
			duration: 100,
			error: { code: 0, description: "OK"},
			nativeEvent: evt,
			paused: false,
			seekable: null,
			seeking: false,
			src: "test"
		};
		const playerState = getObservablePlayerProperties(player, evt);
		expect(playerState).toMatchSnapshot();
	});

	test("getObservablePlayerProperties with bad values", () => {
		const evt = "test";
		const player = {
			buffered: 0,
			currentSrc: "test",
			currentTime: undefined,
			duration: undefined,
			nativeEvent: evt,
			paused: false,
			seekable: null,
			seeking: false,
			src: "test"
		};
		const playerState = getObservablePlayerProperties(player, evt);
		expect(playerState).toMatchSnapshot();
	});
});
