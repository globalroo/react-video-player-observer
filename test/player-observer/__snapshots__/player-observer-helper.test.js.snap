// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getObservableProperties returns expected structure getObservablePlayerProperties 1`] = `
Object {
  "buffered": 0,
  "currentSrc": "test",
  "currentTime": 10,
  "duration": 100,
  "error": Object {
    "code": 0,
    "description": "OK",
  },
  "nativeEvent": "test",
  "paused": false,
  "remainingTime": 90,
  "seekeable": null,
  "seeking": false,
  "src": "test",
}
`;

exports[`getObservableProperties returns expected structure getObservablePlayerProperties with bad values 1`] = `
Object {
  "buffered": 0,
  "currentSrc": "test",
  "currentTime": undefined,
  "duration": undefined,
  "error": Object {
    "code": 0,
    "description": "OK",
  },
  "nativeEvent": "test",
  "paused": false,
  "remainingTime": 0,
  "seekeable": null,
  "seeking": false,
  "src": "test",
}
`;
